src/main/scala/scalaz/parsers/Parsing.scala

```diff
@@ -0,0 +1,287 @@
+package scalaz.parsers
+
+import scalaz.data.~>
+import scalaz.tc.ProfunctorClass.DeriveDimap
+import scalaz.tc._
+
+object Parsing {
+
+  def apply[F[_]: Applicative, G[_]: Applicative](
+    implicit C1_ : Category[λ[(α, β) => α => F[β]]],
+    C2_ : Category[λ[(α, β) => α => G[β]]]
+  ): Parsing[F, G] =
+    new Parsing[F, G] {
+      override val F: Applicative[F]                    = implicitly
+      override val G: Applicative[G]                    = implicitly
+      override val C1: Category[λ[(α, β) => α => F[β]]] = implicitly
+      override val C2: Category[λ[(α, β) => α => G[β]]] = implicitly
+    }
+}
+
+sealed trait Parsing[F[_], G[_]] {
+
+  implicit protected val F: Applicative[F]
+  implicit protected val G: Applicative[G]
+  implicit protected val C1: Category[λ[(α, β) => α => F[β]]]
+  implicit protected val C2: Category[λ[(α, β) => α => G[β]]]
+
+  type Transform[=>:[_, _]] = InstanceOf[TransformClass[=>:]]
+
+  sealed trait TransformClass[=>:[_, _]] extends StrongClass[=>:] with CategoryClass[=>:] {
+
+    def duplicate[A]: A =>: (A, A)
+    def swap[A, B]: (A, B) =>: (B, A)
+    def dropFirst[A, B]: (A, B) =>: B
+
+    def dropSecond[A, B]: (A, B) =>: A =
+      compose[(A, B), (B, A), A](dropFirst, swap)
+
+    def combine[A, B, C](ab: A =>: B, ac: A =>: C): A =>: (B, C) =
+      compose(compose(first[A, B, C](ab), second[A, C, A](ac)), duplicate[A])
+
+    // todo: this is correct but insane
+    def conjunction[A, B, C, D](ab: A =>: B, cd: C =>: D): (A, C) =>: (B, D) = {
+      val x3: (A, C) =>: (C, B)                     = compose(swap, first[A, B, C](ab))
+      val x4: (A, C) =>: (A, D)                     = second[C, D, A](cd)
+      val x5: (A, C) =>: ((C, B), (A, D))           = combine(x3, x4)
+      val x6: ((C, B), (A, D)) =>: (A, ((C, B), D)) = swap2[(C, B), A, D]
+      val x7: (A, ((C, B), D)) =>: (D, (C, B)) =
+        compose(swap[(C, B), D], dropFirst[A, ((C, B), D)])
+      val x8: (D, (C, B)) =>: (C, (D, B)) = swap2[D, C, B]
+      val x9: (C, (D, B)) =>: (B, D)      = compose(swap[D, B], dropFirst[C, (D, B)])
+      val x0: (A, C) =>: (B, D)           = compose(compose(compose(compose(x9, x8), x7), x6), x5)
+      x0
+    }
```

Comment:
So, I had a little fun with this last night 😉 

First, I started by defining a compose operator:

```
implicit class ComposeOp[A, B](val f: A =>: B) {
      def ∘ [C](g: C =>: A): C =>: B = compose(f, g)
    }
```

Then, I was able to abstract out a few interesting operations and gave them names that have some vague association with what they do:

```
def tie[A, B, C, D](ab: A =>: B, cd: C =>: D): (A, C) =>: ((C, B), (A, D)) =
      combine[(A, C), (C, B), (A, D)](compose(swap, first[A, B, C](ab)), second[C, D, A](cd))

def curl[A, B, X]: (X, (A, B)) =>: (B, A) = swap[A, B] ∘ dropFirst[X, (A, B)]
```

Staring at this functions for a bit, one could _almost_ see that might represent something generically useful.  What's fascinating is what happens when we put this all back together and define a more elegant/clear version of `conjunction`:

```
def conjunction_alternate[A, B, C, D](ab: A =>: B, cd: C =>: D): (A, C) =>: (B, D) =
      curl[D, B, C] ∘ swap2[D, C, B] ∘ curl[(C, B), D, A] ∘ swap2[(C, B), A, D] ∘ tie[A, B, C, D](ab, cd)
```

Interesting to see the patterns of the types and the operations?  i.e. How are compositions are alternating between `curl` and `swap2` and how each `curl` has types that correspond with those of a complimentary `swap2`.  I'm not sure what this means, but it's definitely interesting.

Reply:
Thanks @ryanonsrc !
It is definitely better than the ad-hoc version I included. My biggest concern, as we discussed earlier, is the performance comparing to the version implemented with `Applicative`.